src/s3-md5.ts

```diff
@@ -0,0 +1,129 @@
+import { stringifyError } from "./errors.js";
+import * as actionsCore from "@actions/core";
+import { createHash } from "node:crypto";
+import { PathLike } from "node:fs";
+import { FileHandle, open } from "node:fs/promises";
+
+export interface AwsS3Etag {
+  hash: string;
+  chunks: number | undefined;
+}
+
+export function parseEtag(etag: string): AwsS3Etag | undefined {
+  if (!etag.includes("-")) {
+    return {
+      hash: etag,
+      chunks: undefined,
+    };
+  }
+
+  const parts = etag.split("-", 2);
+  if (parts.length !== 2) {
+    actionsCore.debug(
+      `Parsing etag ${etag} failed: expected 2 parts but got ${parts.length}`,
+    );
+    return undefined;
+  }
+
+  const chunks = parseInt(parts[1], 10);
+  if (isNaN(chunks)) {
+    actionsCore.debug(`Parsing etag ${etag} failed: ${parts[1]} is NaN`);
+    return undefined;
+  }
+
+  return {
+    hash: parts[0],
+    chunks,
+  };
+}
+
+export function cleanEtag(inputEtag: string): string {
+  let etag = inputEtag;
+  if (etag.startsWith("W/")) {
+    etag = etag.substring(2);
+  }
+
+  if (etag.startsWith('"') && etag.endsWith('"')) {
+    etag = etag.substring(1, etag.length - 1);
+  }
+
+  return etag;
+}
+
+export async function verifyEtag(
+  filename: PathLike,
+  quotedExpectedEtag: string,
+): Promise<"valid" | "corrupt"> {
+  try {
+    const expectedEtag = cleanEtag(quotedExpectedEtag);
+    const parsedEtag = parseEtag(expectedEtag);
+    if (parsedEtag === undefined) {
+      actionsCore.info(
+        `Verifying etag failed: etag did not parse: ${expectedEtag}`,
+      );
+      return "corrupt";
+    }
+
+    const fd = await open(filename, "r");
+
+    let actualEtag: string;
+    if (parsedEtag.chunks === undefined) {
+      actionsCore.debug(`Verifying etag with a simple md5`);
+      actualEtag = await calculateMd5Etag(fd);
+    } else {
+      actionsCore.debug(`Verifying etag with a chunked md5 from s3`);
+      actualEtag = await calculateS3ChunkedEtag(fd, parsedEtag.chunks);
+    }
+
+    await fd.close();
+
+    if (expectedEtag === actualEtag) {
+      return "valid";
+    } else {
+      actionsCore.info(
+        `Verifying etag failed: etag mismatch. Wanted ${expectedEtag}, got ${actualEtag}`,
+      );
+      return "corrupt";
```

Comment:
I'm not a huge fan of returning strings that we need to compare against... can we rename this to `isEtagValid` and have it instead return a boolean?